pkg/vm/stack.go

```diff
@@ -163,64 +128,43 @@ func NewStack(n string) *Stack {
 
 func newStack(n string, refc *refCounter) *Stack {
 	s := new(Stack)
+	s.elems = make([]Element, 0, 16) // Most of uses are expected to fit into 16 elements.
```

Comment:
I'd very interested to see some statistics around this (and we technically can get for testnet/mainnet), but for now it's just some common sense. Verification contexts are typical to have 1-7 keys with 1-5 signatures (committee 11/21 transactions are rare). Regular executions these days mostly use stack to prepare some parameters for function calls (the data resides in slots otherwise) and these are rarely have more than 5 parameters. So 16 seems to be OK for most of regular use cases.

We just want to reduce the number of micro-allocations here, going from 0 to 1, to 2, to 4, to 8, all of this is not very productive. If the script wants to stress our stacks it still can do this up to expected 2048 elements.

Maybe I'll do some measurements for chains we have though, just as some reality check.

Reply:
It'd be interesting to see the statistics data, yes.

Reply:
There you go. This is stack size got _after_ `vm.execute` for every instruction.

Testnet with 194707 and VerifyTransactions set to true (to run all verifications as well):
```
3233175 stack:  0
6532108 stack:  1
5130095 stack:  2
2909005 stack:  3
2791694 stack:  4
1845768 stack:  5
 513339 stack:  6
 351732 stack:  7
 223084 stack:  8
 195926 stack:  9
 195939 stack:  10
 195858 stack:  11
 195612 stack:  12
 195679 stack:  13
 194934 stack:  14
    294 stack:  15
    219 stack:  16
    100 stack:  17
    154 stack:  18
     45 stack:  19
     44 stack:  20
     44 stack:  21
     44 stack:  22
     44 stack:  23
     44 stack:  24
    210 stack:  25
    150 stack:  26
     22 stack:  27
      6 stack:  28
      6 stack:  29
      6 stack:  30
      6 stack:  31
      6 stack:  32
      6 stack:  33
      6 stack:  34
```
Mainnet:
```
 757120 stack:  0
1132761 stack:  1
 707750 stack:  2
 484875 stack:  3
 318278 stack:  4
 294109 stack:  5
 142513 stack:  6
 123407 stack:  7
 123085 stack:  8
 123076 stack:  9
 123276 stack:  10
 123280 stack:  11
 123011 stack:  12
 123010 stack:  13
 123010 stack:  14
      7 stack:  15
      7 stack:  16
      7 stack:  17
      7 stack:  18
      7 stack:  19
      6 stack:  20
      6 stack:  21
      6 stack:  22
      6 stack:  23
      6 stack:  24
      6 stack:  25
      6 stack:  26
      6 stack:  27
      6 stack:  28
      6 stack:  29
      6 stack:  30
      6 stack:  31
      6 stack:  32
      6 stack:  33
      6 stack:  34
```
I think it confirms 16 as a good number for this.

Reply:
Wow, very precise estimation.